### PR TITLE
Make ed448 compatible with new ed25519 API

### PR DESCRIPTION
**Pull request by armfazh**
Mimics the API of circl/sign/ed25519.
Updates eddilithium4 accordingly.

Merge after #132

### CODE REVIEW

File: .github/workflows/ci-actions.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
     runs-on: ubuntu-18.04
     strategy:
       matrix:
-        GOVER: ['1.14', '1.13', '1.12']
+        GOVER: ['1.14', '1.13']
```
Comment (bwesterb) [inline .github/workflows/ci-actions.yml:15]: Why are we removing support for 1.12?

(Latest Debian is still on 1.11.  Ubuntu 19.10 is on 1.12.  Ubuntu 20.04 is on 1.13.)

File: sign/ed25519/ed25519.go [hunk 2]
```diff
@@ -1,144 +1,217 @@
...
+//  | All above   | (PrivateKey).Sign | VerifyAny     | As above          |
+//
+// Specific functions for sign and verify are defined. A generic signing
+// function for all schemes is available through the crypto.Signer interface,
+// which is implemented by the PrivateKey type. A correspond all-in-one
+// verification method is provided by the VerifyAny function.
+//
+// Signing with Ed25519Ph or Ed25519Ctx requires a context string for domain
+// separation. This parameter is passed using a SignerOptions struct defined
+// in this package.
```
Comment (bwesterb) [inline sign/ed25519/ed25519.go:19]: Perhaps explain the difference between Ed25519Ph and Ed25519Ctx.

File: sign/ed25519/ed25519.go [hunk 3]
```diff
@@ -1,144 +1,217 @@
...
+//
+// Compatibility with crypto.ed25519
+//
+// These functions are compatible with the “Ed25519” function defined in
+// RFC-8032. However, unlike RFC 8032's formulation, this package's private
+// key representation includes a public key suffix to make multiple signing
+// operations with the same key more efficient. This package refers to the
+// RFC-8032 private key as the “seed”.
```
Comment (bwesterb) [inline sign/ed25519/ed25519.go:27]: Perhaps mention that this package can be used as a drop-in replacement of `crypto/ed25519`.

File: sign/ed25519/ed25519.go [hunk 4]
```diff
@@ -1,144 +1,217 @@
...
-func (kp *KeyPair) Sign(rand io.Reader, message []byte, opts crypto.SignerOpts) ([]byte, error) {
+// Use a SignerOptions struct (defined in this package) to pass a context
+// string for signing.
+func (priv PrivateKey) Sign(
+	rand io.Reader,
+	message []byte,
+	opts crypto.SignerOpts) (signature []byte, err error) {
 	var ctx string
-	if o, ok := opts.(*Options); ok {
+	var scheme SchemeID
```
Comment (bwesterb) [inline sign/ed25519/ed25519.go:137]: Perhaps explicitly set or note that the default is Ed25519.

File: sign/ed25519/ed25519_test.go [hunk 5]
```diff
@@ -1,390 +1,89 @@
+// Copyright 2016 The Go Authors. All rights reserved.
```
Comment (bwesterb) [inline sign/ed25519/ed25519_test.go:1]: Is this copyright notice really appropriate?

**Comment by bwesterb** [comment]
@claucece Might want to take a look as well.

File: .github/workflows/ci-actions.yml (hunk 1 above)
Comment (armfazh) [reply .github/workflows/ci-actions.yml:15]: The issue is that crypto/ed25519 appears inside stdlib as of 1.13. 
The public key is an alias of the stdlb crypto/ed25519. This alias plays nicely with type assertions, for example those found in x509.CreateCertificate.
An alternative for not removing support of 1.12 is welcomed.

Comment (bwesterb) [reply .github/workflows/ci-actions.yml:15]: We can control the definition of PublicKey by a build tag.  The "go1.13" build tag is satisfied on Go 1.13 and beyond.

Comment (armfazh) [reply .github/workflows/ci-actions.yml:15]: The issue with that is that golang.org/x/crypto/ed25519 becomes now a dependency, but only for 1.12, and not for 1.13.
how do we express this conditional dependencies in the go.mod file?

**Comment by claucece** [comment]
Oh, nice! I'll take a look later today ;)

File: .github/workflows/ci-actions.yml (hunk 1 above)
Comment (bwesterb) [reply .github/workflows/ci-actions.yml:15]: I meant: use crypto/ed25519.PublicKey for go1.13 and beyond and simply []byte in previous Go versions.  Go 1.12 and before doesn't support Ed25519 in crypto/x509 anyway.

Comment (mmcloughlin) [reply .github/workflows/ci-actions.yml:15]: Go support policy is the last two major releases, so I tend to think it's reasonable for libraries to follow the same rules.

https://golang.org/doc/devel/release.html#policy

As far as I am concerned Debian/Ubuntu are too far behind, and library authors should not have to accommodate them. People should be installing Go some other way on those platforms.

Comment (claucece) [reply .github/workflows/ci-actions.yml:15]: I'm actually pretty much in favor or removing 1.12, it creates some problems as the API changed a lot.

Debian 10 should have go 1.13: https://tecadmin.net/install-go-on-debian/ for what I understand.

**Comment by claucece** [comment]
This looks very nice.. should we do the same for circl/ed25519?

File: .github/workflows/ci-actions.yml (hunk 1 above)
Comment (bwesterb) [reply .github/workflows/ci-actions.yml:15]: The latest in Debian stable (Buster at the moment) is [1.11](https://packages.debian.org/buster/golang).

It's easy to install the latest Go version from source, but that does adds some friction to adoption.

I'm not hell bend on supporting very old version (also as we don't have a big user base yet), but if it's not too much effort, then it does ease adoption.

File: sign/ed25519/ed25519.go [hunk 6]
```diff
@@ -1,144 +1,225 @@
 // Package ed25519 implements Ed25519 signature scheme as described in RFC-8032.
 //
-// References:
-//  - RFC8032 https://rfc-editor.org/rfc/rfc8032.txt
-//  - Ed25519 https://ed25519.cr.yp.to/
-//  - High-speed high-security signatures. https://doi.org/10.1007/s13389-012-0027-1
+// This package implements the three signature variants.
```
Comment (claucece) [inline sign/ed25519/ed25519.go:3]: Just opening...

**Pull request merged**